spec/loophp/collection/Iterator/RandomIteratorSpec.php

```diff
@@ -79,6 +80,30 @@ public function it_can_build_an_iterator_without_a_random_seed()
         }
     }
 
+    public function it_can_get_current()
+    {
+        $this->beConstructedWith(new ArrayIterator(['a']));
+        $this->valid()->shouldReturn(false);
+        $this->shouldThrow(BadMethodCallException::class)->during('current');
```

Comment:
doing this before would've resulted in a `TypeError`

Reply:
I will test during the day, curious to see and understand those changes.

However, maybe we should create a PR only with those changes because they are not related to static analysis but merely to iterators.

Reply:
It's related to both the iterator behaviour and fixing static analysis. Effectively the below error is warning us of a TypeError, which happens if `key` or `current` is called when the key is null (out of bounds).

``` ------ --------------------------------------------------------------------------------------------
  Line   RandomIterator.php
 ------ --------------------------------------------------------------------------------------------
  54     Parameter #1 $key of method loophp\collection\Iterator\RandomIterator<TKey,T of
         string>::getNextItemAtKey() expects int, int|null given.
  61     Parameter #1 $key of method loophp\collection\Iterator\RandomIterator<TKey,T of
         string>::getNextItemAtKey() expects int, int|null given.
```

Example which causes a type error:

```PHP
$it = new ArrayIterator([1]);
$rand = new RandomIterator($it);
$rand->rewind();
$rand->next();

var_dump($rand->key());
```

```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to loophp\collection\Iterator\RandomIterator::getNextItemAtKey() must be of the type int, null given, called in /Users/alex/dev/collections/forked/collection/src/Iterator/RandomIterator.php on line 76 and defined in /Users/alex/dev/collections/forked/collection/src/Iterator/RandomIterator.php:109
```

Another example which causes an infinite loop:

```PHP
$it = new ArrayIterator([1]);
$rand = new RandomIterator($it);

var_dump($rand->current());

```

```
Notice: Undefined offset: 0 in /Users/alex/dev/collections/forked/collection/src/Iterator/RandomIterator.php on line 115
PHP Notice:  Undefined offset: 0 in /Users/alex/dev/collections/forked/collection/src/Iterator/RandomIterator.php on line 115```

Reply:
The `BadMethodCallException` helps protect against those and signals to the caller that they're doing something wrong.

Reply:
Allrighty! Good.